Extend region expression to return points and rays

jzuhone opened the pull request:
This PR extends `ds.r` to allow it to return 0 and 1-dimensional data objects. For example:

```python
p = ds.r[0.1,0.3,0.5]
```
returns a `YTPoint` object, and 

```python
ray = ds.r[(300.0,"kpc"),:,(-200.0,"kpc")]
```
returns a `YTOrthoRay`.

Currently, I'm having trouble getting the rays to work, because for some reason composition of a `YTOrthoRay` with a region does not work (it yields empty arrays when querying fields).

- [x] Fix composition of rays with regions
- [x] Add tests
- [x] Update docs

ngoldbaum commented:
Dunno what's up with the issue you're having. I'd poke at the selection operation, e.g. right here:

https://github.com/yt-project/yt/blob/master/yt/utilities/io_handler.py#L148

(I think)

It would be cool if you could generate a non-ortho ray too somehow.

jzuhone commented:
@ngoldbaum -- it's not clear to me how a non-ortho ray would work here with the slicing syntax. any ideas?

ngoldbaum commented:
@jzuhone it seems to be legal python to do this:

```
In [1]: class MyClass(object):
   ...:     def __getitem__(self, item):
   ...:         print(item)
   ...:

In [2]: obj = MyClass()

In [3]: obj[(0.25, 0.35, 0.8):(0.5, 0.7, 0.6)]
slice((0.25, 0.35, 0.8), (0.5, 0.7, 0.6), None)
```

jzuhone commented:
Ok, I've implemented @ngoldbaum's suggestion. It turns out that `ds.ortho_ray` did not know about unitful quantities for the `coord` parameter or character input for the `axis` parameter, so I've fixed those as well.

jzuhone commented:
Ok, I think that this is ready to go now.

ngoldbaum commented on doc/source/analyzing/objects.rst:
```diff
@@ -214,6 +213,34 @@ domain but centered at 0.5 in code units, you can do:::
+of the ray:::
+
+    start = ((500.0, "kpc"), (0.2, "Mpc"), (100.0, "kpc"))
+    end = ((1.0, "Mpc"), (300.0, "kpc"), (0.0, "kpc"))
```
While this might work, I think mixing units like this in the docs is confusing. Why not just say:

```
from yt.units import kpc
start = [500, 200, 100]*kpc
end = [1000, 300, 0]*kpc
```

Or some other variation on that?

ngoldbaum replied:
Or maybe document this more complicated form after documenting the simpler cases (tuple, list, or array of values assumed to be in code length units and YTArrays with explicit units.

ngoldbaum commented on yt/data_objects/region_expression.py:
```diff
@@ -38,12 +40,19 @@ def __getitem__(self, item):
-            # ds.r[0.1:0.9] where it will be expanded along three dimensions.
-            item = (item, item, item)
+            if iterable(item.start) and len(item.start) == 3 and \
+                iterable(item.stop) and len(item.stop) == 3:
```
You might be able to make this logic a little less tortured by defining a helper function:

```
def is_ray_endpoint(obj):
   return iterable(obj) and len(obj) == 3
```

ngoldbaum commented on doc/source/analyzing/objects.rst:
```diff
@@ -214,6 +213,41 @@ domain but centered at 0.5 in code units, you can do:::
+    ortho_ray = ds.r[(500.0, "kpc"), (200, "kpc"):(300.0, "kpc"), (-2.0, "Mpc")]
+
+To create a ray off-axis, use a single slice between the start and end points
+of the ray:::
```
Can you say this is interpreted in code length units?

MatthewTurk commented on yt/data_objects/region_expression.py:
```diff
@@ -56,20 +64,26 @@ def __getitem__(self, item):
         # coordinate?
 
-        if not all(isinstance(v, slice) for v in item):
+        nslices = sum([isinstance(v, slice) for v in item])
```
minor nit: the `[` and `]` are unnecessary here

MatthewTurk commented on yt/data_objects/region_expression.py:
```diff
@@ -133,3 +147,31 @@ def _create_region(self, bounds_tuple):
+        if axis == 1: coord = [coord[1], coord[0]]
+        source = self._create_region(new_slice)
+        ray = self.ds.ortho_ray(axis, coord, data_source=source)
+        if getattr(new_slice[axis].step, "imag", 0.0) != 0.0:
```
this makes me think we should implement a sampling function at some point so that `imag` on a ray is interesting ...

ngoldbaum replied:
I think me and @lindsayad want to do this for `LinePlot`.

MatthewTurk commented on yt/funcs.py:
```diff
@@ -1141,3 +1141,9 @@ def parse_h5_attr(f, attr):
+    if iterable(v):
+        return len(v)
+    else:
+        return 0
```
could you put a comment here noting that instead of this returning 1 if it's not iterable, it returns 0?  and maybe it should be `False` instead?

jzuhone replied:
I'd rather it always return an integer myself, to be honest.
